clients/lcdproc/disk.c

```diff
@@ -124,32 +137,71 @@ disk_screen(int rep, int display, int *flags_ptr)
 	sock_printf(sock, "widget_set D f 1 2 %i %i %i %i v 12\n", lcd_wid, lcd_hgt, lcd_wid, count);
 	for (i = 0; i < count; i++) {
 		char tmp[lcd_wid + 1];	/* should be large enough */
+		char cap[8] = {'\0'};
+		char dev[256];
+		u_int64_t size = 0;
+		int full = 0;
 
-		if (table[i].dev[0] == '\0')
+		// Eewww, goto... but shy of restructuring everything it's 
+		// the cleanest way to skip the rest of the outer loop. :-/
+		next_mnt:
+
+		if (mnt[i].mpoint[0] == '\0')
 			continue;
 
-		if (i >= num_disks) {	/* Make sure we have enough lines... */
-			sock_printf(sock, "widget_add D s%i string -in f\n", i);
-			sock_printf(sock, "widget_add D h%i hbar -in f\n", i);
+		// Prescreen for entries to ignore.
+		for (int ign_ct=0; ign_ct<ignore_count; ign_ct++) {
+			if (strcmp(mnt[i].mpoint, disk_ignore[ign_ct])==0) {
+				ig_offset++;
+				i++;
+				goto next_mnt; // Behold, the one legitimate use for goto ;-)
+			}
 		}
-		if (lcd_wid >= 20) {	/* 20+x columns */
-			sprintf(tmp, "%-*s %6s E%*sF", dev_wid, table[i].dev, table[i].cap, gauge_wid, "");
+
+		// Parse/format the incoming FS data:
+		int i_widget = i-ig_offset; // Index into the widget array due to "ignore" list
+
+		// Naturally this could be condensed with the line formatting below
+		// but I'm inclined to leave it like this to keep the printf formatting more legible.
+
+		if (strlen(mnt[i].mpoint) > dev_wid)
+			sprintf(dev, "-%s", (mnt[i].mpoint) + (strlen(mnt[i].mpoint) - (dev_wid - 1)));
+		else
+			sprintf(dev, "%s", mnt[i].mpoint);
+
+		full = !mnt[i].blocks ? gauge_scale :
+			gauge_scale * (u_int64_t) (mnt[i].blocks - mnt[i].bfree)
+			/ mnt[i].blocks;
+
+		size = (u_int64_t) mnt[i].bsize * mnt[i].blocks;
+		sprintf_memory(cap, (double) size, 1);
+
+		// Actual display/server output
+
+		if (i_widget >= num_disks) {	/* Make sure we have enough lines... */
+			sock_printf(sock, "widget_add D s%i string -in f\n", i_widget);
+			sock_printf(sock, "widget_add D h%i hbar -in f\n", i_widget);
 		}
-		else {		/* < 20 columns */
-			sprintf(tmp, "%-*s E%*sF", dev_wid, table[i].dev, gauge_wid, "");
+		if (lcd_wid >= 20) {	/* 20+x columns */
+			sprintf(tmp, "%-*s %6s E%*sF", dev_wid, dev, cap, gauge_wid, "");
+		} else {		/* < 20 columns */
+			sprintf(tmp, "%-*s E%*sF", dev_wid, dev, gauge_wid, "");
 		}
-		sock_printf(sock, "widget_set D s%i 1 %i {%s}\n", i, i + 1, tmp);
+		sock_printf(sock, "widget_set D s%i 1 %i {%s}\n", i_widget, i_widget + 1, tmp);
 		sock_printf(sock, "widget_set D h%i %i %i %i\n",
-					i, hbar_pos, i + 1, table[i].full);
+					i_widget, hbar_pos, i_widget + 1, full);
 	}
 
 	/* Now remove extra widgets... */
-	for (; i < num_disks; i++) {
-		sock_printf(sock, "widget_del D s%i\n", i);
-		sock_printf(sock, "widget_del D h%i\n", i);
+	for (int rm = i-ig_offset; rm < num_disks; rm++) {
```

Comment:
Can't your reuse the `i_widget` counter here? Having multiple variables meaning the same thing just makes everything more difficult to comprehend.

Reply:
In this version i_widget was out of scope since it was loop-local. I've reworked this.